### PR TITLE
logstash HA development

### PR DESCRIPTION
**Pull request by maweina**
1. separate logstash into log collector and api collector; occi ensures api collector HA
2. occimon decides api collector leader based on zookeeper; occimon rpm package must be installed on all hosts where api collector are deployed
3. only one api collector is running at any time; if api collector is dead, another api collector must be started

### CODE REVIEW

File: occimon/src/main/java/com/occi/leader/election/Main.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import org.apache.curator.framework.CuratorFrameworkFactory;
+import org.apache.curator.framework.recipes.leader.CancelLeadershipException;
+import org.apache.curator.framework.recipes.leader.LeaderSelector;
+import org.apache.curator.framework.recipes.leader.LeaderSelectorListener;
+import org.apache.curator.framework.recipes.leader.LeaderSelectorListenerAdapter;
+import org.apache.curator.retry.ExponentialBackoffRetry;
+import org.apache.curator.framework.state.ConnectionState;
+
+public class Main {
+    public static String exec[];
```
Comment (zhaoyim) [inline occimon/src/main/java/com/occi/leader/election/Main.java]: why use the un init static var? Do you need the class var? I think this also can be an instance var in line 25 as:
String exec[] = new String[args.length - 2];

File: occimon/src/main/java/com/occi/leader/election/Main.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        LeaderSelectorListener listener = new LeaderSelectorListenerAdapter() {
+            public Process child = null;
+            public void takeLeadership(CuratorFramework client) throws Exception {
+                // this callback will get called when you are the leader
+                // do whatever leader work you need to and only exit
+                // this method when you want to relinquish leadership
+                System.out.println("I'm the leader now ");
+                try {
+                    System.out.println("Starting child");
+                    child = Runtime.getRuntime().exec(exec);
```
Comment (zhaoyim) [inline occimon/src/main/java/com/occi/leader/election/Main.java]: One question here, we use the monitor to start the logstash daemon, if the monitor process stoped, seems the logstash daemon will still running. We checked this in the Ambari scripts?

File: occimon/src/main/java/com/occi/leader/election/Main.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        RetryPolicy retryPolicy = new ExponentialBackoffRetry(1000, 3);
+        CuratorFramework client = CuratorFrameworkFactory.newClient(zookeeperConnectionString, retryPolicy);
+
+        LeaderSelectorListener listener = new LeaderSelectorListenerAdapter() {
+            public Process child = null;
+            public void takeLeadership(CuratorFramework client) throws Exception {
+                // this callback will get called when you are the leader
+                // do whatever leader work you need to and only exit
+                // this method when you want to relinquish leadership
+                System.out.println("I'm the leader now ");
```
Comment (zhaoyim) [inline occimon/src/main/java/com/occi/leader/election/Main.java]: System out and err will we add them into the log? Now just print

File: src/package/scripts/logstash_ha.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+"""
+from resource_management import *
+import sys
+from copy import deepcopy
+
+def logstash_ha(role=None):
```
Comment (qianguo1) [inline src/package/scripts/logstash_ha.py]: the "logstash_ha" in python code need to be renamed as "metric collector"

**Pull request merged**